packages/frames.js/src/middleware/farcasterHubContext.ts

```diff
@@ -0,0 +1,88 @@
+import {
+  FrameActionPayload,
+  FrameMessageReturnType,
+  getFrameMessage,
+} from "..";
+import {
+  InvalidFrameActionPayloadError,
+  RequestBodyNotJSONError,
+} from "../core/errors";
+import { FramesMiddleware, JsonValue } from "../core/types";
+
+function isValidFrameActionPayload(
+  value: unknown
+): value is FrameActionPayload {
+  return (
+    typeof value === "object" &&
+    !!value &&
+    "trustedData" in value &&
+    "untrustedData" in value
+  );
+}
+
+async function decodeFrameActionPayloadFromRequest(
+  request: Request
+): Promise<FrameActionPayload | undefined> {
+  try {
+    // use clone just in case someone wants to read body somewhere along the way
+    const body = await request
+      .clone()
+      .json()
+      .catch(() => {
+        throw new RequestBodyNotJSONError();
+      });
+
+    if (!isValidFrameActionPayload(body)) {
+      throw new InvalidFrameActionPayloadError();
+    }
+
+    return body;
+  } catch (e) {
+    if (
+      e instanceof RequestBodyNotJSONError ||
+      e instanceof InvalidFrameActionPayloadError
+    ) {
+      return undefined;
+    }
+
+    console.error(e);
+
+    return undefined;
+  }
+}
+
+type FrameMessage = Omit<
+  FrameMessageReturnType<{ fetchHubContext: true }>,
+  "message"
+> & { state?: JsonValue };
+
+type FramesMessageContext = {
+  message?: FrameMessage;
+};
+
+export function farcasterHubContext(): FramesMiddleware<
+  any,
+  FramesMessageContext
+> {
+  return async (context, next) => {
+    // frame message is available only if the request is a POST request
+    if (context.request.method !== "POST") {
+      return next();
+    }
+
+    // body must be a JSON object
+    const payload = await decodeFrameActionPayloadFromRequest(context.request);
+
+    if (!payload) {
+      return next();
+    }
+
+    const message = await getFrameMessage(payload, {
+      fetchHubContext: true,
```

Comment:
we should also provide a way to override the hub url for debugger to work right? That could be made a part of options and be available on `FramesContext.options` or something like that.

Reply:
ah yes good point